Added autoclicker, better mouse, Vim like editing layer and more.

yuannan opened the pull request:
<!--- Provide a general summary of your changes in the title above. -->
A new layout for the DZ65RGB v3 that has a much wider feature set.

## Description
Mouse now has more button than ever, including all 4 directional scrolls and mouse buttons 1-5.

Including an autoclicker in 5 speeds. The amount of speeds and their values are easily configurable. This includes a toggleable clicker that can change speeds, a burst in that speed and a quick burst hold down to click as quick as the board can output them.

FN layers reflect that of vim and vimlike keybinds for faster navigation using the FN layer and home row.

For those who wishes to use a GUI VIA/VIAL support is also included.
<!--- Describe your changes in detail here. -->

## Types of Changes

<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply. -->
- [ ] Core
- [ ] Bugfix
- [X] New feature
- [X] Enhancement/optimization
- [X] Keyboard (addition or update)
- [X] Keymap/layout/userspace (addition or update)
- [ ] Documentation


## Checklist

<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [X] My code follows the code style of this project: [**C**](https://docs.qmk.fm/#/coding_conventions_c), [**Python**](https://docs.qmk.fm/#/coding_conventions_python)
- [X] I have read the [**PR Checklist** document](https://docs.qmk.fm/#/pr_checklist) and have made the appropriate changes.
- [ ] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [X] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [ ] I have added tests to cover my changes.
- [X] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

drashna commented on keyboards/dztech/dz65rgb/keymaps/yuannan/keymap.c:
```diff
@@ -0,0 +1,119 @@
+#include QMK_KEYBOARD_H
```
Could you add a GPL2+ compatible license header here? 

For instance:
```c
// Copyright 2021 Your Name (@yourgithub)
// SPDX-License-Identifier: GPL-2.0-or-later
```

Or 

https://github.com/qmk/qmk_firmware/blob/6b299fab6e6a3143c061b9ea005253361fcd524b/data/templates/base/keyboard.c#L1-L15

drashna commented on keyboards/dztech/dz65rgb/keymaps/yuannan/keymap.c:
```diff
@@ -0,0 +1,119 @@
+    if (autoclick_state) {
+      mousekey_off(KC_MS_BTN1);
+    } else {
+      mousekey_on(KC_MS_BTN1);
```
You could use `tap_code(KC_BTN1)` rather than the mousekey functions.

yuannan replied:
Orignally it had a release timer built in as well. But I scrapped that. Thanks for your help!

drashna commented on keyboards/dztech/dz65rgb/keymaps/yuannan/yuannan.vil:
```diff
@@ -0,0 +1 @@
+{"version": 1, "uid": -1, "layout": [[["KC_GESC", "KC_1", "KC_2", "KC_3", "KC_4", "KC_5", "KC_6", "KC_7", "KC_8", "KC_9", "KC_0", "KC_MINUS", "KC_EQUAL", "KC_BSPACE", "KC_HOME"], ["KC_TAB", "KC_Q", "KC_W", "KC_E", "KC_R", "KC_T", "KC_Y", "KC_U", "KC_I", "KC_O", "KC_P", "KC_LBRACKET", "KC_RBRACKET", "KC_BSLASH", "KC_PGUP"], ["KC_LCTRL", "KC_A", "KC_S", "KC_D", "KC_F", "KC_G", "KC_H", "KC_J", "KC_K ...
```
Vial files are not supported or used in the main repo and need to be removed. 
```suggestion
```

yuannan commented:
I've updated the repo with the appropriate files @drashna. I've also updated it with a auto-clicker work around in Linux as libinput by default enables a debounce feature preventing the autoclicker from working properly.

My apologies for not reading the docs properly but I hope my actions have been rectified. Please make me aware of any other issues with my current code.

After a few nights of experimentation and debugging was I was able to create a .bin of "25910/26624 (97%, 714 bytes free)" Lower than Windows. so if any future users experience sizes over the current maximum it may be worth compiling it on Linux.

drashna commented on keyboards/dztech/dz65rgb/keymaps/yuannan/README.md:
```diff
@@ -0,0 +1,70 @@
+	- Deleting entire line
+	- Deleting from cursor to start of line
+	- Deleting form cursor to end of line
+6. VIA/VIAL support.
```
Vial isn't supported in the repo, so we generally remove references to it to avoid potential confusion. 
```suggestion
6. VIA support.
```

drashna commented on keyboards/dztech/dz65rgb/keymaps/yuannan/keymap.c:
```diff
@@ -0,0 +1,204 @@
+*/
+
+#include QMK_KEYBOARD_H
+#include "mousekey.h"
```
```suggestion
```

drashna commented:
Also, just a heads up, it's a newer function, but you may want to check out the deferred executors stuff: 
https://docs.qmk.fm/#/custom_quantum_functions?id=deferred-execution

If you need a reference for how to implement it, you can see what I've done in `/users/drashna/oled/oled_stuff.c`, though that's a basic example.  You'd want to use the "cancel" function to stop the auto-click.

fauxpark commented on keyboards/dztech/dz65rgb/keymaps/yuannan/README.md:
```diff
@@ -0,0 +1,70 @@
+[DZ65RGB/v3 Yuannan's Layout Imgur Album](https://imgur.com/a/BI2RjgE)
```
Please rename this file to lowercase.

yuannan commented:
> Also, just a heads up, it's a newer function, but you may want to check out the deferred executors stuff: https://docs.qmk.fm/#/custom_quantum_functions?id=deferred-execution
> 
> If you need a reference for how to implement it, you can see what I've done in `/users/drashna/oled/oled_stuff.c`, though that's a basic example. You'd want to use the "cancel" function to stop the auto-click.

Thank you guys for the great work on the new API. The CPS numbers I'm seeing are now very consistent and clean regardless of other settings. This seem to cut down task interference quite a lot. The space complexity could be better with it taking about ~450B more compared to my old. There were other changes made so that number could be vastly larger or greater.

Special thanks to @drashna  for all the helpful tips you've given me and the example code. There is a typo on line 900 (https://github.com/qmk/qmk_firmware/blob/935881cca9ef5fb1a2cdea7543ca57a772781781/users/drashna/oled/oled_stuff.c#L900).

It should be "trigger" instead of "triger" with 2 gs.

drashna commented on keyboards/dztech/dz65rgb/keymaps/yuannan/config.h:
```diff
@@ -0,0 +1,44 @@
+#define RGB_MATRIX_VAL_STEP 8
+#define RGB_MATRIX_SPD_STEP 8
+
+#define USB_POLLING_INTERVAL_MS 1
```
This won't be needed in a week, as it is to be the new default at that time. 

```suggestion
```

yuannan commented:
> Thank you for your contribution! This pull request has been automatically marked as stale because it has not had activity in the last 45 days. It will be closed in 30 days if no further activity occurs. Please feel free to give a status update now, or re-open when it's ready. For maintainers: Please label with `awaiting review`, `breaking_change`, `in progress`, or `on hold` to prevent the issue from being re-flagged.

As far as I'm aware I've done everything that has been asked of me. Just don't have perms to merge.

drashna commented:
> As far as I'm aware I've done everything that has been asked of me. Just don't have perms to merge.

No worries.  It's more of an issue with stale-bot being ...

yuannan commented:
> > As far as I'm aware I've done everything that has been asked of me. Just don't have perms to merge.
> 
> No worries. It's more of an issue with stale-bot being ...

Is there anything I should do? This pull seem to be open for quite a while. I'm happy to fix something I've done or create new tasks.

keyboard-magpie commented:
For future reference, we recommend against committing to your `master` branch as you've done here, because pull requests from modified `master` branches can make it more difficult to keep your QMK fork updated. It is highly recommended for QMK development – regardless of what is being done or where – to keep your master updated, but **NEVER** commit to it. Instead, do all your changes in a branch (branches are basically free in Git) and issue PRs from your branches when you're developing.

There are instructions on how to keep your fork updated here:

[**Best Practices: Your Fork's Master: Update Often, Commit Never**](https://docs.qmk.fm/#/newbs_git_using_your_master_branch)

[Fixing Your Branch](https://docs.qmk.fm/#/newbs_git_resynchronize_a_branch) will walk you through fixing up your `master` branch moving forward. If you need any help with this just ask.

Thanks for contributing!
